src/PPU.cpp

```diff
@@ -98,122 +69,273 @@ void PPU::updateLY() {
-		for (int j = 0; j < 8; j++) {
-			uint8_t lo = mmu->read(start + (index % 128) * 16 + j * 2);
-			uint8_t hi = mmu->read(start + (index % 128) * 16 + j * 2 + 1);
+		// Determine the index of the tile in the data, and determine the correct
+		// starting location for block of tile data given that index
+		uint8_t bi = mmu->directRead(backgroundStart + i);
+		uint16_t bs = (bi > 127) ? secondHalfStart : firstHalfStart;
+		bi %= 128;
 
-			set_line(x * 8, j + y * 8, hi, lo);
-		}
+		uint8_t wi = mmu->directRead(windowStart + i);
+		uint16_t ws = (wi > 127) ? secondHalfStart : firstHalfStart;
+		wi %= 128;
+
+		setTile(backgroundBuffer, MAP_WIDTH, bs, bi, x * 8, y * 8, bgp);
+		setTile(windowBuffer, MAP_WIDTH, ws, wi, x * 8, y * 8, bgp);
 	}
 }
 
-uint8_t PPU::getSCY() {
-	return mmu->directRead(0xFF42);
+void PPU::updateObjects() {
+	// Early out if LCD is off
+	bool lcdc7 = mmu->directRead(0xFF40) & (1 << 7);
+	if (!lcdc7) {
+		return;
+	}
+
+	// Update palette information
+	uint32_t obp0Data = mmu->directRead(0xFF48);
+	uint32_t obp1Data = mmu->directRead(0xFF49);
+
+	uint32_t obp0[] = {
+		0,
+		(obp0Data & 0x0C) >> 2,
+		(obp0Data & 0x30) >> 4,
+		(obp0Data & 0xC0) >> 6
+	};
+
+	uint32_t obp1[] = {
+		0,
+		(obp1Data & 0x0C) >> 2,
+		(obp1Data & 0x30) >> 4,
+		(obp1Data & 0xC0) >> 6
+	};
+
+	uint16_t oamStart = 0xFE00;
+	uint16_t tileStart = 0x8000;
+
+	// Clear objects array of old sprites
+	std::fill(objectsBuffer, objectsBuffer + MAP_WIDTH * MAP_HEIGHT, 0);
+
+	// Iterate 4 bytes at a time from 0xFE00 to 0xFE9F
+	for (int i = 0; i < 40; i++) {
+		uint8_t y = mmu->directRead(oamStart + i * 4);
+		uint8_t x = mmu->directRead(oamStart + i * 4 + 1);
+		uint8_t tileIndex = mmu->directRead(oamStart + i * 4 + 2);
+		uint8_t flags = mmu->directRead(oamStart + i * 4 + 3);
+
+		uint32_t* obp = (flags & (1 << 4)) ? obp1 : obp0;
+		uint32_t yFlip = (flags & (1 << 6));
+		uint32_t xFlip = (flags & (1 << 5));
+
+		setObject(objectsBuffer, MAP_WIDTH, tileStart, tileIndex, x, y, obp, xFlip, yFlip);
+	}
 }
 
-uint8_t PPU::getSCX() {
-	return mmu->directRead(0xFF43);
+void PPU::updateScanline() {
+	// Early out if scanline is greater than 143, shouldn't ever hit this point
+	if (scanline > 143) {
+		return;
+	}
+
+	// If lcdc5 = 1, window is enabled
+	bool lcdc5 = mmu->directRead(0xFF40) & (1 << 5);
+
+	uint8_t scx = mmu->directRead(0xFF43);
+	uint8_t scy = mmu->directRead(0xFF42);
+	uint8_t wx = mmu->directRead(0xFF4B);
+	uint8_t wy = mmu->directRead(0xFF4A);
+
+	// Iterate through every pixel on the current scanline and set to the correct value from the
+	// background, window, and objects buffers.
+	for (int i = 0; i < 160; i++) {
+		uint8_t x = i;
+		uint8_t y = scanline;
+
+		uint16_t si = y * 160 + x;
+
+		// Get the current position of the background tilemap to set in the screen,
+		// including wrapping around if it would exceed the bounds of the tilemap.
+		// TODO: figure out at what point constants are less clear than the actual number
```

Comment:
I mention constants here, I think at some point I should do a pass and figure out where constants make sense and where hard coded values make sense. I think it makes sense for some things in the game boy should just be hard coded.